File: crypto.py
import hashlib
import hmac
import base64
import json
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad, unpad

class CryptoEngine:
    def __init__(self):
        # AES ব্লকের সাইজ ১৬ বাইট
        self.block_size = 16

    def generate_hmac(self, key_k1, message):
        """মেসেজ M এবং কি K1 ব্যবহার করে F1/F2 তৈরি করা"""
        byte_key = key_k1.encode('utf-8')
        byte_message = message.encode('utf-8')
        h = hmac.new(byte_key, byte_message, hashlib.sha256)
        return h.hexdigest()

    def _derive_aes_key(self, password_k2, fingerprint_bp, timestamp_t):
        """K2, BP এবং T থেকে একটি ३२-বাইটের AES কী তৈরি করা"""
        combined = f"{password_k2}{fingerprint_bp}{timestamp_t}"
        return hashlib.sha256(combined.encode()).digest()

    def encrypt_data(self, message, f1, k2, bp, t):
        """M + F1 কে K2, BP, T দিয়ে এনক্রিপ্ট করা"""
        combined_data = json.dumps({"M": message, "F1": f1})
        key = self._derive_aes_key(k2, bp, t)
        
        cipher = AES.new(key, AES.MODE_CBC)
        iv = cipher.iv # ইনপুট ভেক্টর
        
        ct_bytes = cipher.encrypt(pad(combined_data.encode(), self.block_size))
        
        # IV এবং সাইফারটেক্সট একত্রে বেস৬৪ করে পাঠানো হয়
        return base64.b64encode(iv + ct_bytes).decode('utf-8')

    def decrypt_data(self, encrypted_data, k2, bp, t):
        """Ciphertext ডিক্রিপ্ট করে M এবং F1 উদ্ধার করা"""
        try:
            data = base64.b64decode(encrypted_data)
            key = self._derive_aes_key(k2, bp, t)
            
            iv = data[:16]
            ct = data[16:]
            
            cipher = AES.new(key, AES.MODE_CBC, iv)
            pt_bytes = unpad(cipher.decrypt(ct), self.block_size)
            
            return json.loads(pt_bytes.decode('utf-8'))
        except Exception as e:
            return None

File: src/utils/crypto.ts
// Frontend crypto utility to match backend crypto.py
import CryptoJS from 'crypto-js';

export class FrontendCryptoEngine {
  private blockSize = 16;

  /**
   * Generate HMAC using K1 and message
   */
  generateHmac(keyK1: string, message: string): string {
    return CryptoJS.HmacSHA256(message, keyK1).toString();
  }

  /**
   * Derive AES key from K2, BP, and T
   */
  private deriveAesKey(k2: string, bp: string, t: string): string {
    const combined = `${k2}${bp}${t}`;
    return CryptoJS.SHA256(combined).toString();
  }

  /**
   * Encrypt data (M + F1) using K2, BP, T
   */
  encryptData(message: string, f1: string, k2: string, bp: string, t: string): string {
    try {
      const combinedData = JSON.stringify({ M: message, F1: f1 });
      const key = this.deriveAesKey(k2, bp, t);

      // Convert hex key to WordArray for CryptoJS
      const keyArray = CryptoJS.enc.Hex.parse(key);
      
      // Generate a random IV
      const iv = CryptoJS.lib.WordArray.random(128/8);
      
      // Encrypt with AES-CBC
      const encrypted = CryptoJS.AES.encrypt(combinedData, keyArray, {
        mode: CryptoJS.mode.CBC,
        iv: iv,
        padding: CryptoJS.pad.Pkcs7,
      });

      // Concatenate IV + ciphertext and encode as base64 to match Python backend format
      // encrypted.ciphertext is the actual cipher bytes (before base64 encoding)
      const ciphertextWords = encrypted.ciphertext;
      const combined = iv.concat(ciphertextWords);
      return CryptoJS.enc.Base64.stringify(combined);
    } catch (error) {
      console.error('Encryption error details:', {
        message,
        f1Length: f1.length,
        k2Length: k2.length,
        bpLength: bp.length,
        tLength: t.length,
        error
      });
      throw error;
    }
  }

  /**
   * Decrypt ciphertext to get M and F1
   */
  decryptData(encryptedData: string, k2: string, bp: string, t: string): { M: string; F1: string } | null {
    try {
      const key = this.deriveAesKey(k2, bp, t);
      const keyArray = CryptoJS.enc.Hex.parse(key);

      const decrypted = CryptoJS.AES.decrypt(encryptedData, keyArray, {
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7,
      });

      const decryptedString = decrypted.toString(CryptoJS.enc.Utf8);
      return JSON.parse(decryptedString);
    } catch (error) {
      console.error('Decryption error:', error);
      return null;
    }
  }
}

export const cryptoEngine = new FrontendCryptoEngine();
